### PR TITLE
Create ast iterator and optimize global variable access

### PR DESCRIPTION
**Pull request by hugomg**

### CODE REVIEW

File: spec/global_upvalues_spec.lua [hunk 1]
```diff
@@ -0,0 +1,85 @@
+local global_upvalues = require "titan-compiler.global_upvalues"
+
+local function run_global_upvalues(code)
+    local prog, errs = global_upvalues.analyze("(global_upvalues_spec)", code)
+    return prog, table.concat(errs, "\n")
+end
+
+describe("Global variable pass:", function()
+
```
Comment (gligneul) [inline spec/global_upvalues_spec.lua]: s/\n//

File: titan-compiler/ast_iterator.lua [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+--  customize it by overriding some of the methods.
+function ast_iterator.new()
+    local iterator = {}
+    for name, func in pairs(ast_iterator.default) do
+        iterator[name] = func
+    end
+    return iterator
+end
+
+
```
Comment (gligneul) [inline titan-compiler/ast_iterator.lua]: s/\n//

File: titan-compiler/ast_iterator.lua [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+        self:Exp(var.exp, ...)
+
+    else
+        error("impossible")
+    end
+end
+
+function ast_iterator.default:Exp(exp, ...)
+    local tag = exp._tag
+    if     tag == ast.Exp.Nil then
```
Comment (gligneul) [inline titan-compiler/ast_iterator.lua]: We should either have the `-- Nothing to do` in the same line here like the `Type` method.

File: titan-compiler/scope_analysis.lua [hunk 4]
```diff
@@ -69,31 +62,17 @@ end
 -- bind_names
 --
 
-bind_names_program = function(prog, st, errors)
+
```
Comment (gligneul) [inline titan-compiler/scope_analysis.lua]: s/\n//

File: titan-compiler/scope_analysis.lua [hunk 5]
```diff
@@ -103,30 +82,19 @@ bind_names_type = function(type_node, st, errors)
...
-        for _, ptype in ipairs(type_node.argtypes) do
-            bind_names_type(ptype, st, errors)
-        end
-        for _, rettype in ipairs(type_node.rettypes) do
-            bind_names_type(rettype, st, errors)
-        end
-
     else
-        error("impossible")
+        ast_iterator.default.Type(self, type_node, st, errors)
```
Comment (gligneul) [inline titan-compiler/scope_analysis.lua:85]: Why not use the `:` notation here?

File: titan-compiler/global_upvalues.lua [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+--     In Toplevel.Func nodes.
+--     List of integers, describes what global variables the function uses.
+function global_upvalues.analyze(filename, input)
+    local prog, errors = checker.check(filename, input)
+    if not prog then return false, errors end
+    analyze:Program(prog)
+    return prog, errors
+end
+
+
```
Comment (gligneul) [inline titan-compiler/global_upvalues.lua]: s/\n//

File: titan-compiler/global_upvalues.lua [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+            tlnode._global_index = n_globals
+            n_globals = n_globals + 1
+        end
+    end
+    prog._n_globals = n_globals
+
+    ast_iterator.default.Program(self, prog)
+end
+
+
```
Comment (gligneul) [inline titan-compiler/global_upvalues.lua]: s/\n//

File: titan-compiler/global_upvalues.lua [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+    if     tag == ast.Toplevel.Func then
+        local referenced_globals_map = {}
+        analyze:Stat(tlnode.block, referenced_globals_map)
+        local referenced_globals = {}
+        for index, _ in pairs(referenced_globals_map) do
+            table.insert(referenced_globals, index)
+        end
+        table.sort(referenced_globals)
+        tlnode._referenced_globals = referenced_globals
+
```
Comment (gligneul) [inline titan-compiler/global_upvalues.lua]: s/\n//

File: spec/global_upvalues_spec.lua [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+        local atzero = prog[2]
+        assert.equals(1, atzero._global_index)
+        assert.equals(0, #atzero._referenced_globals)
+        local f = prog[3]
+        assert.equals(2, f._global_index)
+        assert.equals(1, #f._referenced_globals)
+        assert.equals(0, f._referenced_globals[1])
+    end)
+end)
+
```
Comment (gligneul) [inline spec/global_upvalues_spec.lua]: s/\n//

**Review by gligneul — CHANGES_REQUESTED** [review]
Minor stuff

File: titan-compiler/scope_analysis.lua (hunk 5 above)
Comment (hugomg) [reply titan-compiler/scope_analysis.lua:85]: Because we need to explicitly pass the `self`

**Review by gligneul — APPROVED** [review]

**Pull request merged**